### ISSUE TITLE
Bug: [prefer-nullish-coalescing] should consider the difference between null and undefined with strict equality

### ISSUE THREAD

**Issue by DanieleGrizzi** [typescript-eslint/typescript-eslint#9181]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.5&fileType=.tsx&code=FAYw9gdgzgLgBAMzGAXHWAnAlhA5nAHzggFcAbMwuEiAEwFMEd7a4Beaux52gbmGBIw7NhxoMmEFnAD8cAORC4WKJwk95cNEN5A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WlqYSNkAC1pkA9gEMkyMswDm6KL2jjokcGAC%2BILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const foo: string | null | undefined = undefined;

foo === undefined ? 'foo is undefined' : foo;
```

### ESLint Config

```javascript
{
  "root": true,
  "overrides": [
    {
      "files": ["*.ts"],
      "parserOptions": {
        "project": ["**/tsconfig.json"],
        "createDefaultProgram": true
      },
      "extends": [
        "plugin:@typescript-eslint/recommended",
        "plugin:@typescript-eslint/recommended-type-checked",
        "plugin:@typescript-eslint/strict",
        "plugin:@typescript-eslint/strict-type-checked",
        "plugin:@typescript-eslint/stylistic",
        "plugin:@typescript-eslint/stylistic-type-checked"
      ]
    }
  ]
}
```

### tsconfig

```jsonc
{
  "compileOnSave": false,
  "compilerOptions": {
    "outDir": "./dist/out-tsc",
    "strict": true,
    "noImplicitOverride": true,
    "noPropertyAccessFromIndexSignature": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "sourceMap": true,
    "declaration": false,
    "experimentalDecorators": true,
    "moduleResolution": "bundler",
    "importHelpers": true,
    "target": "ES2022",
    "module": "ES2022",
    "useDefineForClassFields": false,
    "lib": ["ES2022", "dom"],
    "noUncheckedIndexedAccess": true,
    "resolveJsonModule": true,
  },
  "angularCompilerOptions": {
    "enableI18nLegacyMessageIdFormat": false,
    "strictInjectionParameters": true,
    "strictInputAccessModifiers": true,
    "strictTemplates": true
  }
}
```

### Expected Result

eslint should not recognize it as an error and should not suggest to use coaleshing operator (??) cause if you use it the result will be different.

Example:

ORIGINAL
const foo: string | null | undefined = undefined;
foo === undefined ? 'foo is undefined' : foo;
The result is 'foo is undefined' only if foo is 'undefined'

ESLINT FIX
const foo: string | null | undefined = null;
foo ?? 'foo is undefined';
The result is 'foo is undefined' even when foo is 'null'

### Actual Result

eslint recognize it as an error

**Comment by auvred** [comment]
The problem is that the type of `foo` is always `undefined`. TypeScript doesn't tell us that `foo` may be a `null`, because it knows that it's const and initialized as `undefined`. For more information, see this sample: 


```ts
const foo: string | null | undefined = undefined;

foo === undefined ? 'foo is undefined' : foo;
//^?
// foo is always undefined



const bar: string | null | undefined = foo as string | null | undefined;

bar === undefined ? 'bar is undefined' : bar;
//^?
// bar is string | null | undefined
```

[playground link](https://typescript-eslint.io/play/#ts=5.4.5&fileType=.tsx&code=MYewdgzgLgBAZiEAuG0BOBLMBzGAfGMAVwBsT8YiwATAUzi1upgF5Kb7HqBuAKF4QhWLNlToMwTGAH4YAckEwMEduK5yYKQXwD0OgHrTee%2BIiUqAhiQDuFgJ4qxnSdX79QkWACMLaFOixcAmIyCicJKTZFCxUAnAoQ8gJwrj5eHzRhUQ4I5lk5DPNVZyYNFAzdAyMTQuVUKEx44NIk4tygA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WlqYSNkAC1pkA9gEMkyMswDm6KL2jjokcGAC%2BILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by DanieleGrizzi** [comment]
Thank you! :)
